### PR TITLE
[SPARK-46285][SQL] Add `foreachWithSubqueries`

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

We can have a `foreachWithSubqueries` which also traverse the subqueries in the query plan.

### Why are the changes needed?

Add a new way to access subqueries in the query plan.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

UT

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Comment by amaliujia** [comment]
@cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -4709,6 +4710,17 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
     val df6 = df3.join(df2, col("df3.zaak_id") === col("df2.customer_id"), "outer")
     df5.crossJoin(df6)
   }
+
+  test("SPARK-46285: foreachWithSubqueries") {
+    val sql = "SELECT * FROM t WHERE col_t in (SELECT * FROM subquery_table)"
+    val plan = spark.sessionState.sqlParser.parsePlan(sql)
+    val cache = scala.collection.mutable.Set[String]()
+    plan.foreachWithSubqueries {
```
[the remark this answers is not in the document]
Comment (amaliujia) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: done

**Comment by amaliujia** [comment]
@cloud-fan I updated the test case. Please take a look.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/LogicalPlanSuite.scala [hunk 2]
```diff
@@ -145,4 +146,16 @@ class LogicalPlanSuite extends SparkFunSuite {
     assert(query.where(Literal.FalseLiteral).maxRows.contains(0))
     assert(query.where(Literal.FalseLiteral).maxRowsPerPartition.contains(0))
   }
+
+  test("SPARK-46285: foreachWithSubqueries") {
+    val input = UnresolvedRelation(Seq("subquery_table"))
+    val input2 = UnresolvedRelation(Seq("t"))
+    val plan = Project(Seq(Alias(ScalarSubquery(input), "s")()), input2)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/LogicalPlanSuite.scala:153]: Super nit: this plan is actually invalid as `ScalarSubquery` requires the plan to return at least one row. Maybe `Exists` is better for testing, we can use it in `Filter`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/LogicalPlanSuite.scala [hunk 3]
```diff
@@ -145,4 +146,16 @@ class LogicalPlanSuite extends SparkFunSuite {
...
+    val cache = scala.collection.mutable.Set[String]()
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/LogicalPlanSuite.scala:154]:
```suggestion
    val tableNames = scala.collection.mutable.Set[String]()
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/LogicalPlanSuite.scala [hunk 4]
```diff
@@ -145,4 +146,16 @@ class LogicalPlanSuite extends SparkFunSuite {
...
+    plan.foreachWithSubqueries {
+      case UnresolvedRelation(iden, _, _) => cache.add(iden.mkString("."))
```
Comment (beliefer) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/LogicalPlanSuite.scala:156]: How about using `name`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/LogicalPlanSuite.scala (hunk 2 above)
Comment (amaliujia) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/LogicalPlanSuite.scala]: Dot it. Done.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/LogicalPlanSuite.scala (hunk 4 above)
Comment (amaliujia) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/LogicalPlanSuite.scala]: done

Comment (beliefer) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/LogicalPlanSuite.scala]: What I mean is using
```
  /** Returns a `.` separated name for this relation. */
  def tableName: String = multipartIdentifier.quoted

  override def name: String = tableName
```

Comment (amaliujia) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/LogicalPlanSuite.scala]: ah I see :) done.

**Review by cloud-fan — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Thank you, @amaliujia , @cloud-fan , @beliefer .
Merged to master for Apache Spark 4.0.0.

**Review by beliefer — COMMENTED** [review]
Late LGTM.

**Pull request closed without merging**